### PR TITLE
Space preferences for whether or not you see DMs in a Space

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-web/issues/19529
Fixes https://github.com/vector-im/element-web/issues/19955

image
image

---
Here's what your changelog entry will look like:

## ✨ Features
 * Space preferences for whether or not you see DMs in a Space ([\#7250](https://github.com/matrix-org/matrix-react-sdk/pull/7250)). Fixes vector-im/element-web#19529 and vector-im/element-web#19955.

Preview: https://61bc56fb8e98931c57f57f19--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by andybalaam** [comment]
Looks good to me, but too many things I don't yet understand to approve it.

**Comment by andybalaam** [comment]
I tested this using netlify and it appears to work.  I can enable a new "People" section in the Spaces bar and it appears to work.

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
Works well, thanks.

Just want to amend the copy, please:

> This groups your chats with members of this space.  Turning this off will hide those chats from your view of Space name.

**Review by niquewoodhouse — APPROVED** [review]

File: src/stores/spaces/SpaceStore.ts [hunk 1]
```diff
@@ -349,176 +350,335 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
+            switch (s.getMyMembership()) {
+                case "join":
+                    arr[0].push(s);
+                    break;
+                case "invite":
+                    arr[1].push(s);
+                    break;
+            }
+            return arr;
+        }, [[], []] as [Room[], Room[]]);
```
Comment (turt2live) [inline src/stores/spaces/SpaceStore.ts:462]: This bit takes a few reads to really understand what it's doing, but overall is fine. Suggestion would be to leverage `getEffectiveMembership` even though it doesn't return anything more helpful - just makes it a bit clearer what the overall intent is down the road. Adding names to the intermediate array also helps a bit.

Something like this:

```suggestion
        const [joinedSpaces, invitedSpaces] = visibleSpaces.reduce(([joined, invited], s) => {
            const membership = getEffectiveMembership(s.getMyMembership);
            switch (membership) {
                case EffectiveMembership.Join:
                    joined.push(s);
                    break;
                case EffectiveMembership.Invite:
                    invited.push(s);
                    break;
            }
            return [joined, invited];
        }, [[], []] as [Room[], Room[]]);
```

File: src/stores/spaces/SpaceStore.ts [hunk 2]
```diff
@@ -349,176 +350,335 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
+        // Home is effectively a super set of People + Orphans with the addition of having all invites too.
+        if (enabledMetaSpaces.has(MetaSpace.Orphans) || enabledMetaSpaces.has(MetaSpace.Home)) {
             const orphans = visibleRooms.filter(r => {
                 // filter out DMs and rooms with >0 parents
                 return !this.parentMap.get(r.roomId)?.size && !DMRoomMap.shared().getUserIdForRoomId(r.roomId);
             });
             this.spaceFilteredRooms.set(MetaSpace.Orphans, new Set(orphans.map(r => r.roomId)));
         }
 
+        if (this.activeSpace[0] !== "!") {
```
Comment (turt2live) [inline src/stores/spaces/SpaceStore.ts:538]: This needs a comment - what does `!` denote?

Comment (turt2live) [reply src/stores/spaces/SpaceStore.ts:538]: actually, is this getting the first character of the potential room ID? Can we not do that please?

File: src/stores/spaces/SpaceStore.ts [hunk 3]
```diff
@@ -901,18 +1133,39 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
-                            if (this.activeSpace[0] !== "!" &&
-                                !enabledMetaSpaces.includes(this.activeSpace as MetaSpace)
-                            ) {
-                                this.goToFirstSpace();
+                            const hasPeopleOrHomeEnabled = this.enabledMetaSpaces.some(s => {
+                                return s === MetaSpace.Home || s === MetaSpace.People;
+                            });
+
+                            // if a metaspace currently being viewed was removed, go to another one
+                            if (this.activeSpace[0] !== "!" && !newValue[this.activeSpace]) {
```
Comment (turt2live) [inline src/stores/spaces/SpaceStore.ts:1145]: `!` behaviour reliance comment applies here too

File: src/utils/sets.ts [hunk 4]
```diff
@@ -32,3 +34,16 @@ export function setHasDiff<T>(a: Set<T>, b: Set<T>): boolean {
         return true; // different lengths means they are naturally diverged
     }
 }
+
+type Diff<T> = { added: T[], removed: T[] };
```
Comment (turt2live) [inline src/utils/sets.ts:38]: can we raise this type so `arrayDiff` can use it too?

or, better yet, move it to the arrays.ts and export it from there. Then return it and have the `setDiff` for sets return `keyDiff` as-is, using the imported type.

**Review by turt2live — CHANGES_REQUESTED** [review]
otherwise seems fine - just a couple code quality things to consider.

File: src/stores/spaces/SpaceStore.ts (hunk 2 above)
Comment (t3chguy) [reply src/stores/spaces/SpaceStore.ts:538]: its either a room ID if a real space is selected or an enum member if a meta space is selected

Comment (turt2live) [reply src/stores/spaces/SpaceStore.ts:538]: Really not comfortable with this sort of check for whether or not something references a room :( 

`!matrixClient.getRoom(this.activeSpace)` would be better, but also not great. It should be just as fast for the purposes of the functions here.

Comment (t3chguy) [reply src/stores/spaces/SpaceStore.ts:538]: will make a utility function which explicitly checks against all enum members instead, as a room ID of a room your client doesn't know about might be valid one day

File: src/utils/arrays.ts [hunk 5]
```diff
@@ -184,6 +184,8 @@ export function arrayHasDiff(a: any[], b: any[]): boolean {
     }
 }
 
+export type Diff<T> = { added: T[], removed: T[] };
```
Comment (turt2live) [inline src/utils/arrays.ts:187]: changing the return type of `arrayDiff` would be great 😇

File: src/utils/sets.ts [hunk 6]
```diff
@@ -32,3 +34,14 @@ export function setHasDiff<T>(a: Set<T>, b: Set<T>): boolean {
...
+/**
+ * Determines the values added and removed between two sets.
+ * @param a The first set. Must be defined.
+ * @param b The second set. Must be defined.
+ * @returns The difference between the values in each set.
+ */
+export function setDiff<T>(a: Set<T>, b: Set<T>): Diff<T> {
+    const keyDiff = arrayDiff(Array.from(a), Array.from(b));
+    return { added: keyDiff.added, removed: keyDiff.removed };
+}
```
Comment (turt2live) [inline src/utils/sets.ts:47]: if the return type of `arrayDiff` gets changed, this becomes:
```suggestion
export function setDiff<T>(a: Set<T>, b: Set<T>): Diff<T> {
    return arrayDiff(Array.from(a), Array.from(b));
}
```

**Review by turt2live — APPROVED** [review]
lgtm either way, though I'd love to see the type fixes

File: src/utils/arrays.ts (hunk 5 above)
Comment (t3chguy) [reply src/utils/arrays.ts:187]: I must've been half asleep when doing this ;(

**Pull request merged**